File: app/src/core/scoring.ts
/** Exploratory score — şeffaf ağırlıklar. Terapötik/metafizik kanıt DEĞİLDİR. */
import { analyzeFrequency } from "./math";
import type { CorpusNumber, EvidenceLevel } from "../data/types";

const CENTS_TIGHT = 8;
const CENTS_LOOSE = 25;
const EVIDENCE_PENALTY: Record<EvidenceLevel, number> = { direct: 0, derived: 5, interpretive: 15, speculative: 30 };
const ORDER: EvidenceLevel[] = ["direct", "derived", "interpretive", "speculative"];

export interface Score {
  hz: number;
  math: number;
  historical: number;
  exploratory: number;
  corpusHits: CorpusNumber[];
  experimental: number | null;
  replication: number | null;
}

export const mathematicalScore = (hz: number): number => {
  const a = analyzeFrequency(hz);
  let s = 0;
  if (a.integer) s += 5;
  if (a.factorization.length) s += Math.max(0, 10 - 2 * new Set(a.factorization).size);
  const best = a.ratios[0].errorCents;
  if (best <= CENTS_TIGHT) s += 25;
  else if (best <= CENTS_LOOSE) s += 10;
  return Math.min(s, 40);
};

export const historicalScore = (sourceCount: number, directNumeric: boolean, level: EvidenceLevel): number =>
  Math.max(0, Math.min(30, sourceCount * 5) + (directNumeric ? 10 : 0) - EVIDENCE_PENALTY[level]);

export const scoreFrequency = (hz: number, corpus: CorpusNumber[]): Score => {
  const hits = Number.isInteger(hz) ? corpus.filter((n) => n.value === hz) : [];
  let historical = 0;
  if (hits.length) {
    const best = hits.reduce((b, h) => (ORDER.indexOf(h.evidenceLevel) < ORDER.indexOf(b.evidenceLevel) ? h : b));
    historical = historicalScore(new Set(hits.map((h) => h.sourceId)).size, hits.some((h) => h.extraction === "explicit"), best.evidenceLevel);
  }
  const math = mathematicalScore(hz);
  return { hz, math, historical, exploratory: math + historical, corpusHits: hits, experimental: null, replication: null };
};

File: engine/aurora/scoring.py
"""Exploratory score — şeffaf, önceden kaydedilebilir ağırlıklar.

UYARI: Bu puan terapötik/metafizik etki kanıtı DEĞİLDİR. Sadece araştırma adaylarını sıralar.
Experimental + Replication olmadan hiçbir aday 'validated' olamaz.
"""
from __future__ import annotations

from .corpus import Corpus
from .math_analysis import analyze_frequency

CENTS_TIGHT, CENTS_LOOSE = 8.0, 25.0
EVIDENCE_PENALTY = {"direct": 0.0, "derived": 5.0, "interpretive": 15.0, "speculative": 30.0}
WARNING = "Exploratory ranking; not evidence of therapeutic or metaphysical efficacy."


def mathematical_score(freq: float) -> float:
    a = analyze_frequency(freq)
    score = 0.0
    if a.integer:
        score += 5.0                                   # yeniden üretilebilirlik (küçük katkı)
    if a.factorization:
        score += max(0.0, 10.0 - 2.0 * len(set(a.factorization)))   # düşük karmaşıklık
    best = a.notable_ratios[0]
    if best.error_cents <= CENTS_TIGHT:
        score += 25.0
    elif best.error_cents <= CENTS_LOOSE:
        score += 10.0
    return min(score, 40.0)


def historical_score(source_count: int, direct_numeric: bool, evidence_level: str) -> float:
    base = min(30.0, source_count * 5.0) + (10.0 if direct_numeric else 0.0)
    return max(0.0, base - EVIDENCE_PENALTY.get(evidence_level, 30.0))


def historical_from_corpus(freq: float, corpus: Corpus) -> tuple[float, list[dict]]:
    """Frekans tam sayıysa corpus'ta aynı sayıyı arar; en iyi kanıt seviyesiyle puanlar."""
    hits = corpus.numbers_equal(freq)
    if not hits:
        return 0.0, []
    order = ["direct", "derived", "interpretive", "speculative"]
    best = min(hits, key=lambda h: order.index(h["evidence_level"]))
    direct = any(h["extraction"] == "explicit" for h in hits)
    return historical_score(len({h["source_id"] for h in hits}), direct, best["evidence_level"]), hits


def candidate_score(freq: float, corpus: Corpus | None = None, source_count: int = 0,
                    direct_numeric: bool = False, evidence_level: str = "speculative") -> dict:
    math_s = mathematical_score(freq)
    hits: list[dict] = []
    if corpus is not None:
        hist_s, hits = historical_from_corpus(freq, corpus)
    else:
        hist_s = historical_score(source_count, direct_numeric, evidence_level)
    return {
        "frequency_hz": freq,
        "mathematical_score": round(math_s, 3),
        "historical_score": round(hist_s, 3),
        "exploratory_score": round(math_s + hist_s, 3),
        "corpus_hits": [h["number_id"] for h in hits],
        "experimental_score": None,
        "replication_score": None,
        "warning": WARNING,
    }


def rank_candidates(freqs: list[float], corpus: Corpus | None = None) -> list[dict]:
    return sorted((candidate_score(f, corpus) for f in freqs),
                  key=lambda r: r["exploratory_score"], reverse=True)
